src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
The recent change is slightly different than what was discussed here.  All of the recent work hid the transact API from the caller, which meant a caller could not create their own complete transaction implementation and/or could not guarantee that this code would not call a read method if the caller required that guarantee.  The latest change reverted modifications made to the transact methods and introduced a transactFully method.  The transactFully method returns an InputStream that handles the full read and implements a close method that flushes the remainder of the transaction.